File: app/iosApp/iosApp/NativeModules/AppRouteHandler.swift
import UIKit

enum NativeRouteAnimation: String {
  case standard = "default"
  case fade
  case none
}

enum NativeRoutePresentation: String {
  case page
  case inputDialog
  case overlay
}

private func withRouteFadeTransition(on layer: CALayer, action: () -> Void) {
  let transition = CATransition()
  transition.duration = 0.25
  transition.type = .fade
  layer.add(transition, forKey: "lynx.route.fade")
  action()
}

private struct ValidatedRoute {
  let bundle: String
  let animation: NativeRouteAnimation
  let presentation: NativeRoutePresentation
  let statusBarStyle: NativeStatusBarStyle
  let params: [String: Any]
  let presentScrimColor: UIColor?
  let presentBackdropTransition: Bool
  let presentEnterAnimation: PresentContentAnimationOptions
  let presentExitAnimation: PresentContentAnimationOptions
  let presentBackdropBlur: Bool
  let presentIOSSwipeDown: Bool
  let presentDragDownToDismiss: Bool

  var routeData: [String: Any] {
    [
      "bundle": bundle,
      "statusBarStyle": statusBarStyle.rawValue,
      "animation": animation.rawValue,
      "presentation": presentation.rawValue,
      "params": params,
    ]
  }
}

/// A pending openForResult callback, resolved when the opened route dies.
private final class PendingRouteResult {
  let onResult: LynxCallbackBlock
  weak var opener: UIViewController?
  var result: [String: Any]?

  init(onResult: @escaping LynxCallbackBlock, opener: UIViewController?) {
    self.onResult = onResult
    self.opener = opener
  }
}

/// Host navigation behind the autolinked Router module: opens another Lynx
/// bundle by pushing a LynxPageViewController from the calling host. The
/// module resolves the host from its own Lynx view, so the handler stays
/// stateless. `presentation: 'overlay'` snapshots the calling page first; the
/// pushed page replays that snapshot as its backdrop (see PresentBackdrop)
/// instead of using a modal transparent presentation, so the iOS-like present
/// transition also composes with the navigation back stack.
///
/// openForResult routes carry a result token on the pushed page; the pending
/// callback registry here delivers the recorded result (or none) from the
/// page's actual stack removal, which fires on every close path — close(),
/// closeWithResult(), the interactive pop gesture and the present dismiss
/// choreography alike.
final class AppRouteHandler: NSObject, LynxRouteHandler {
  // Main-thread confined: written from DispatchQueue.main blocks only.
  private static var pendingResults: [String: PendingRouteResult] = [:]
  private static var nextToken = 0

  // The ObjC selector `openFromViewController:options:success:` imports
  // into Swift as `open(from:options:success:)`.
  func open(
    from host: UIViewController,
    options: [String: Any],
    success completion: @escaping LynxCallbackBlock
  ) {
    let parsed = parseRoute(options)
    guard let route = parsed.route else {
      completion(parsed.error ?? "Invalid route options")
      return
    }
    DispatchQueue.main.async {
      guard host.viewIfLoaded?.window != nil else {
        completion("Router has no visible UIViewController host")
        return
      }
      guard self.openRoute(route, from: host, resultToken: nil) else {
        completion("Router cannot open a route from the current host")
        return
      }
      completion("")
    }
  }

  func openForResult(
    from host: UIViewController,
    options: [String: Any],
    onResult: @escaping LynxCallbackBlock
  ) {
    let parsed = parseRoute(options)
    guard let route = parsed.route else {
      onResult(Self.errorEnvelope(parsed.error ?? "Invalid route options"))
      return
    }
    DispatchQueue.main.async {
      Self.nextToken += 1
      let token = "route-result-\(Self.nextToken)"
      Self.pruneDeadOpeners()
      Self.pendingResults[token] = PendingRouteResult(onResult: onResult, opener: host)
      guard self.openRoute(route, from: host, resultToken: token) else {
        Self.pendingResults.removeValue(forKey: token)
        onResult(Self.errorEnvelope("Router cannot open a route from the current host"))
        return
      }
    }
  }

  func close(
    from host: UIViewController,
    success completion: @escaping LynxCallbackBlock
  ) {
    DispatchQueue.main.async {
      self.performClose(from: host, completion: completion)
    }
  }

  func closeWithResult(
    from host: UIViewController,
    result: [String: Any],
    success completion: @escaping LynxCallbackBlock
  ) {
    DispatchQueue.main.async {
      if let page = host as? LynxPageViewController, let token = page.routeResultToken {
        Self.pendingResults[token]?.result = result
      }
      self.performClose(from: host, completion: completion)
    }
  }

  /// Delivers the pending entry once, with its recorded result or none; safe
  /// from any thread (delivery itself stays on the main thread).
  static func completeRouteResult(token: String) {
    if Thread.isMainThread {
      deliverRouteResult(token: token)
    } else {
      DispatchQueue.main.async {
        Self.deliverRouteResult(token: token)
      }
    }
  }

  // MARK: Internals

  private func performClose(
    from host: UIViewController,
    completion: @escaping LynxCallbackBlock
  ) {
    guard let page = host as? LynxPageViewController else {
      completion("The root route cannot be closed")
      return
    }
    if page.isInputDialogRoute, page.presentingViewController != nil {
      page.requestDialogDismiss {
        page.dismiss(animated: false) {
          completion("")
        }
      }
      return
    }
    guard let navigation = page.navigationController,
          navigation.viewControllers.first !== page else {
      completion("The root route cannot be closed")
      return
    }
    if page.isOverlayRoute {
      // The reverse choreography pops the page itself once the backdrop has
      // restored the previous page's pixels.
      page.playDismissChoreography {
        navigation.popViewController(animated: false)
        completion("")
      }
    } else if page.routeAnimation == .fade {
      withRouteFadeTransition(on: navigation.view.layer) {
        navigation.popViewController(animated: false)
      }
      completion("")
    } else {
      navigation.popViewController(animated: page.routeAnimation != .none)
      completion("")
    }
  }

  /// Opens the validated route in either the navigation stack or its own
  /// keyboard-isolated modal overlay.
  private func openRoute(
    _ route: ValidatedRoute,
    from host: UIViewController,
    resultToken: String?
  ) -> Bool {
    guard host.viewIfLoaded?.window != nil else {
      return false
    }
    if route.presentation == .inputDialog {
      let page = makePage(route, snapshot: nil, resultToken: resultToken)
      page.modalPresentationStyle = .overFullScreen
      page.modalTransitionStyle = .crossDissolve
      page.modalPresentationCapturesStatusBarAppearance = true
      page.isModalInPresentation = true
      host.present(page, animated: false)
      return true
    }
    guard let navigation = host.navigationController else { return false }
    // Preserve an overlay page's currently resolved screen-concentric radius
    // before the nested push changes its navigation-container geometry.
    (host as? LynxPageViewController)?.preservePresentBackdropCornerClipping()
    navigation.setNavigationBarHidden(true, animated: false)
    let snapshot = route.presentation == .overlay
      ? PresentBackdrop.capture(of: host.view, blurred: route.presentBackdropBlur)
      : nil
    let page = makePage(route, snapshot: snapshot, resultToken: resultToken)
    switch route.animation {
    case .standard:
      navigation.pushViewController(page, animated: true)
    case .fade:
      withRouteFadeTransition(on: navigation.view.layer) {
        navigation.pushViewController(page, animated: false)
      }
    case .none:
      navigation.pushViewController(page, animated: false)
    }
    return true
  }

  private func makePage(
    _ route: ValidatedRoute,
    snapshot: UIImage?,
    resultToken: String?
  ) -> LynxPageViewController {
    LynxPageViewController(
      bundleName: route.bundle,
      route: route.routeData,
      snapshot: snapshot,
      statusBarStyle: route.statusBarStyle,
      presentScrimColor: route.presentScrimColor,
      presentBackdropTransition: route.presentBackdropTransition,
      presentEnterAnimation: route.presentEnterAnimation,
      presentExitAnimation: route.presentExitAnimation,
      presentIOSSwipeDownEnabled: route.presentIOSSwipeDown,
      presentDragDownToDismissEnabled: route.presentDragDownToDismiss,
      routeResultToken: resultToken
    )
  }

  private func parseRoute(_ options: [String: Any]) -> (route: ValidatedRoute?, error: String?) {
    guard let bundle = options["bundle"] as? String,
          bundle.range(of: "^[a-z0-9][a-z0-9-]*$", options: .regularExpression) != nil else {
      return (nil, "Invalid Lynx bundle name")
    }

    let rawPresentation = options["presentation"] as? String
      ?? NativeRoutePresentation.page.rawValue
    guard let presentation = NativeRoutePresentation(rawValue: rawPresentation) else {
      return (nil, "Invalid route presentation: \(rawPresentation)")
    }
    let rawAnimation = options["animation"] as? String
      ?? (presentation == .page
        ? NativeRouteAnimation.standard.rawValue
        : NativeRouteAnimation.none.rawValue)
    // Overlay routes own their open/close choreography, so they always run
    // without a system transition regardless of the animation value.
    let animation: NativeRouteAnimation
    if presentation == .overlay {
      animation = .none
    } else if let parsed = NativeRouteAnimation(rawValue: rawAnimation) {
      animation = parsed
    } else {
      return (nil, "Invalid route animation: \(rawAnimation)")
    }
    let rawStatusBarStyle = options["statusBarStyle"] as? String
      ?? NativeStatusBarStyle.darkContent.rawValue
    guard let statusBarStyle = NativeStatusBarStyle(rawValue: rawStatusBarStyle) else {
      return (nil, "Invalid status bar style: \(rawStatusBarStyle)")
    }
    let params = options["params"] as? [String: Any] ?? [:]
    let overlay = options["overlay"] as? [String: Any]
    var presentScrimColor: UIColor? = nil
    if let rawScrimColor = overlay?["scrimColor"] as? String {
      guard let parsed = UIColor(lynxHexARGB: rawScrimColor) else {
        return (nil, "Invalid overlay scrim color: \(rawScrimColor)")
      }
      presentScrimColor = parsed
    }
    let presentBackdropTransition = overlay?["backdropTransition"] as? Bool ?? true
    let legacyContentTransition = overlay?["contentTransition"] as? Bool ?? true
    let presentEnter = overlay?["enter"] as? [String: Any]
    let presentExit = overlay?["exit"] as? [String: Any]
    let presentEnterAnimation = PresentContentAnimationOptions(
      opacity: presentEnter?["opacity"] as? Bool ?? false,
      push: presentEnter?["push"] as? Bool ?? legacyContentTransition
    )
    let presentExitAnimation = PresentContentAnimationOptions(
      opacity: presentExit?["opacity"] as? Bool ?? false,
      push: presentExit?["push"] as? Bool ?? legacyContentTransition
    )
    let presentBackdropBlur = overlay?["backdropBlur"] as? Bool ?? false
    let presentIOSSwipeDown = overlay?["iosSwipeDown"] as? Bool ?? false
    let presentDragDownToDismiss = overlay?["dragDownToDismiss"] as? Bool ?? false
    return (
      ValidatedRoute(
        bundle: bundle,
        animation: animation,
        presentation: presentation,
        statusBarStyle: statusBarStyle,
        params: params,
        presentScrimColor: presentScrimColor,
        presentBackdropTransition: presentBackdropTransition,
        presentEnterAnimation: presentEnterAnimation,
        presentExitAnimation: presentExitAnimation,
        presentBackdropBlur: presentBackdropBlur,
        presentIOSSwipeDown: presentIOSSwipeDown,
        presentDragDownToDismiss: presentDragDownToDismiss
      ),
      nil
    )
  }

  private static func deliverRouteResult(token: String) {
    guard let pending = pendingResults.removeValue(forKey: token),
          pending.opener != nil else {
      return
    }
    pending.onResult(pending.result.map(valueEnvelope) ?? "{}")
  }

  /// Drops entries whose awaiting page died; their JS promise can no longer run.
  private static func pruneDeadOpeners() {
    pendingResults = pendingResults.filter { $0.value.opener != nil }
  }

  private static func errorEnvelope(_ message: String) -> String {
    guard let data = try? JSONSerialization.data(withJSONObject: ["error": message]),
          let json = String(data: data, encoding: .utf8) else {
      return "{}"
    }
    return json
  }

  private static func valueEnvelope(_ result: [String: Any]) -> String {
    guard JSONSerialization.isValidJSONObject(result),
          let data = try? JSONSerialization.data(withJSONObject: ["value": result]),
          let json = String(data: data, encoding: .utf8) else {
      return "{}"
    }
    return json
  }
}

File: app/androidApp/app/src/main/java/com/lynxapp/nativemodule/AppRouteHandler.kt
package com.lynxapp.nativemodule

import android.app.Activity
import android.app.Application
import android.content.Intent
import android.util.Log
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.ReadableMap
import com.lynxapp.LynxTemplateApplication
import com.lynxapp.R
import com.lynxapp.activity.LynxDialogActivity
import com.lynxapp.activity.LynxPageActivity
import com.lynxapp.activity.PresentBackdrop
import com.lynxapp.activity.PresentContentAnimationOptions
import com.lynxapp.activity.RouteSnapshotStore
import com.lynxapp.autolink.navigation.LynxRouteHandler
import com.lynxapp.autolink.navigation.LynxRouteHandler.RouteResultEnvelope
import com.lynxapp.autolink.navigation.NavigationModule.ANIMATION_DEFAULT
import com.lynxapp.autolink.navigation.NavigationModule.ANIMATION_FADE
import com.lynxapp.autolink.navigation.NavigationModule.ANIMATION_NONE
import com.lynxapp.autolink.navigation.NavigationModule.isLynxRouteAnimation
import com.lynxapp.autolink.device.DeviceSystemUI
import com.lynxapp.component.LoadingOverlay
import org.json.JSONObject
import java.lang.ref.WeakReference
import java.util.concurrent.atomic.AtomicLong

private data class PresentRouteOptions(
    val scrimColor: String?,
    val backdropTransition: Boolean,
    val enterAnimation: PresentContentAnimationOptions,
    val exitAnimation: PresentContentAnimationOptions,
    val backdropBlur: Boolean,
    val androidPredictiveBackDown: Boolean,
    val dragDownToDismiss: Boolean,
)

private data class ValidatedRoute(
    val bundle: String,
    val animation: String,
    val presentation: String,
    val statusBarStyle: String,
    val params: Map<String, Any>,
    val presentOptions: PresentRouteOptions,
)

/** A pending openForResult callback, resolved when the opened route dies. */
private class PendingRouteResult(
    val resultCallback: Callback,
    opener: Activity,
) {
    val opener = WeakReference(opener)
    var result: Map<String, Any>? = null
}

/**
 * Host navigation behind the autolinked Navigation module: opens another Lynx
 * bundle in a real Android Activity. When the Application-prefetched OTA
 * version list marks the target bundle outdated, the update downloads behind
 * a loading overlay before the page opens against the fresh cache. State on
 * the calling Activity's route (such as its animation) drives how close()
 * undoes the transition. `presentation: 'overlay'` snapshots the calling page
 * first; the opened page replays that snapshot as its backdrop (see
 * PresentBackdrop) instead of relying on a translucent activity, so the
 * iOS-like present transition also composes with the normal back stack.
 *
 * openForResult routes carry a result token through the launched intent; the
 * pending callback registry here delivers the recorded result (or none) from
 * onActivityDestroyed, which fires on every close path — close(),
 * closeWithResult(), system Back and finishWithPresentTransition alike.
 */
class AppRouteHandler(
    application: Application,
) : LynxRouteHandler, Application.ActivityLifecycleCallbacks {
    init {
        application.registerActivityLifecycleCallbacks(this)
    }

    // Main-thread confined: written from runOnUiThread blocks, read from the
    // registered ActivityLifecycleCallbacks.
    private val pendingResults = HashMap<String, PendingRouteResult>()

    override fun open(activity: Activity, options: ReadableMap, callback: Callback) {
        val route = parseRouteOptions(options, callback::invoke) ?: return
        activity.runOnUiThread {
            runCatching {
                openWhenReady(activity, route, resultToken = null)
            }.onSuccess {
                callback.invoke("")
            }.onFailure { error ->
                callback.invoke(error.message ?: "Unable to open native route")
            }
        }
    }

    override fun openForResult(
        activity: Activity,
        options: ReadableMap,
        resultCallback: Callback,
    ) {
        // A callable reference to the member extension invokeEnvelopeError is
        // prohibited in Kotlin; wrap the call instead.
        val route = parseRouteOptions(options) {
            resultCallback.invokeEnvelopeError(it)
        } ?: return
        val token = "route-result-${NEXT_RESULT_TOKEN.incrementAndGet()}"
        activity.runOnUiThread {
            // Registry access stays on the main thread: openForResult puts,
            // closeWithResult records, onActivityDestroyed delivers.
            pendingResults[token] = PendingRouteResult(resultCallback, activity)
            runCatching {
                openWhenReady(activity, route, resultToken = token)
            }.onFailure { error ->
                pendingResults.remove(token)
                resultCallback.invokeEnvelopeError(
                    error.message ?: "Unable to open native route",
                )
            }
        }
    }

    override fun close(activity: Activity, callback: Callback) {
        if (activity !is LynxPageActivity) {
            callback.invoke("The route cannot be closed")
            return
        }
        activity.runOnUiThread { performClose(activity, callback) }
    }

    override fun closeWithResult(
        activity: Activity,
        result: ReadableMap,
        callback: Callback,
    ) {
        if (activity !is LynxPageActivity) {
            callback.invoke("The route cannot be closed")
            return
        }
        activity.runOnUiThread {
            routeResultToken(activity)?.let { token ->
                pendingResults[token]?.result = result.toHashMap()
            }
            performClose(activity, callback)
        }
    }

    private fun performClose(activity: LynxPageActivity, callback: Callback) {
        if (activity.isRootRoute) {
            // Closing the root leaves the app; the task stays alive so
            // returning is instant instead of a cold Lynx restart.
            activity.moveTaskToBack(true)
        } else if (activity.isOverlayRoute) {
            // The reverse choreography finishes the activity itself once
            // the backdrop has restored the previous page's pixels.
            activity.finishWithPresentTransition()
        } else {
            activity.finish()
            when (activity.routeAnimation) {
                ANIMATION_NONE ->
                    activity.overridePendingTransition(0, 0)
                ANIMATION_FADE ->
                    activity.overridePendingTransition(
                        android.R.anim.fade_in,
                        android.R.anim.fade_out,
                    )
            }
        }
        callback.invoke("")
    }

    /**
     * Opens the route immediately, or once a pending bundle update has
     * downloaded behind [LoadingOverlay] (a failed download opens the page
     * with the current source instead of blocking navigation). The open
     * callback resolves as soon as the flow is initiated.
     */
    private fun openWhenReady(
        activity: Activity,
        route: ValidatedRoute,
        resultToken: String?,
    ) {
        val repository = (activity.application as LynxTemplateApplication).bundleRepository
        val update = repository.pendingUpdateFor(route.bundle)
        if (update == null) {
            startActivityForRoute(activity, route, resultToken)
            return
        }
        LoadingOverlay.show(activity, activity.getString(R.string.updating_bundle))
        repository.download(update) { updated ->
            if (!updated) {
                Log.w(TAG, "Bundle ${route.bundle} update failed; opening the current source")
            }
            activity.runOnUiThread {
                LoadingOverlay.hide(activity)
                if (activity.isFinishing || activity.isDestroyed) return@runOnUiThread
                startActivityForRoute(activity, route, resultToken)
            }
        }
    }

    private fun startActivityForRoute(
        activity: Activity,
        route: ValidatedRoute,
        resultToken: String?,
    ) {
        if (route.presentation == LynxPageActivity.PRESENTATION_OVERLAY) {
            // Capture only after any update overlay has been hidden, so the
            // snapshot shows the page the user was looking at.
            PresentBackdrop.capture(activity, route.presentOptions.backdropBlur) { snapshot ->
                activity.runOnUiThread {
                    if (activity.isFinishing || activity.isDestroyed) {
                        snapshot?.recycle()
                        return@runOnUiThread
                    }
                    if (snapshot != null) {
                        RouteSnapshotStore.put(snapshot)
                    }
                    startRouteActivity(activity, route, resultToken, snapshot != null)
                }
            }
        } else {
            startRouteActivity(activity, route, resultToken, false)
        }
    }

    private fun startRouteActivity(
        activity: Activity,
        route: ValidatedRoute,
        resultToken: String?,
        withSnapshot: Boolean,
    ) {
        val destination = if (
            route.presentation == LynxPageActivity.PRESENTATION_INPUT_DIALOG
        ) {
            LynxDialogActivity::class.java
        } else {
            LynxPageActivity::class.java
        }
        activity.startActivity(
            Intent(activity, destination).apply {
                putExtra(LynxPageActivity.EXTRA_BUNDLE, route.bundle)
                putExtra(LynxPageActivity.EXTRA_ANIMATION, route.animation)
                putExtra(LynxPageActivity.EXTRA_PRESENTATION, route.presentation)
                putExtra(LynxPageActivity.EXTRA_SNAPSHOT, withSnapshot)
                putExtra(LynxPageActivity.EXTRA_STATUS_BAR_STYLE, route.statusBarStyle)
                if (route.presentOptions.scrimColor != null) {
                    putExtra(
                        LynxPageActivity.EXTRA_PRESENT_SCRIM_COLOR,
                        route.presentOptions.scrimColor,
                    )
                }
                putExtra(
                    LynxPageActivity.EXTRA_PRESENT_BACKDROP_TRANSITION,
                    route.presentOptions.backdropTransition,
                )
                putExtra(
                    LynxPageActivity.EXTRA_PRESENT_ENTER_OPACITY,
                    route.presentOptions.enterAnimation.opacity,
                )
                putExtra(
                    LynxPageActivity.EXTRA_PRESENT_ENTER_PUSH,
                    route.presentOptions.enterAnimation.push,
                )
                putExtra(
                    LynxPageActivity.EXTRA_PRESENT_EXIT_OPACITY,
                    route.presentOptions.exitAnimation.opacity,
                )
                putExtra(
                    LynxPageActivity.EXTRA_PRESENT_EXIT_PUSH,
                    route.presentOptions.exitAnimation.push,
                )
                putExtra(
                    LynxPageActivity.EXTRA_PRESENT_BACKDROP_BLUR,
                    route.presentOptions.backdropBlur,
                )
                putExtra(
                    LynxPageActivity.EXTRA_PRESENT_ANDROID_PREDICTIVE_BACK_DOWN,
                    route.presentOptions.androidPredictiveBackDown,
                )
                putExtra(
                    LynxPageActivity.EXTRA_PRESENT_DRAG_DOWN_TO_DISMISS,
                    route.presentOptions.dragDownToDismiss,
                )
                putExtra(
                    LynxPageActivity.EXTRA_PARAMS_JSON,
                    JSONObject(route.params).toString(),
                )
                if (resultToken != null) {
                    putExtra(EXTRA_RESULT_TOKEN, resultToken)
                }
            },
        )
        when (route.animation) {
            ANIMATION_NONE ->
                activity.overridePendingTransition(0, 0)
            ANIMATION_FADE ->
                activity.overridePendingTransition(
                    android.R.anim.fade_in,
                    android.R.anim.fade_out,
                )
        }
    }

    private fun parseRouteOptions(
        options: ReadableMap,
        onError: (String) -> Unit,
    ): ValidatedRoute? {
        val bundle = options.getString("bundle", "")
        if (!BUNDLE_NAME.matches(bundle)) {
            onError("Invalid Lynx bundle name: $bundle")
            return null
        }

        val presentation = options.getString(
            "presentation",
            LynxPageActivity.PRESENTATION_PAGE,
        )
        if (!LynxPageActivity.isRoutePresentation(presentation)) {
            onError("Invalid route presentation: $presentation")
            return null
        }
        val defaultAnimation = if (
            presentation == LynxPageActivity.PRESENTATION_PAGE
        ) {
            ANIMATION_DEFAULT
        } else {
            ANIMATION_NONE
        }
        // Overlay routes own their open/close choreography, so they always
        // run without a system transition regardless of the animation value.
        val animation = if (
            presentation == LynxPageActivity.PRESENTATION_OVERLAY
        ) {
            ANIMATION_NONE
        } else {
            options.getString("animation", defaultAnimation)
        }
        if (!isLynxRouteAnimation(animation)) {
            onError("Invalid route animation: $animation")
            return null
        }
        val statusBarStyle = options.getString(
            "statusBarStyle",
            DeviceSystemUI.STATUS_BAR_STYLE_DARK_CONTENT,
        )
        if (!DeviceSystemUI.isStatusBarStyle(statusBarStyle)) {
            onError("Invalid status bar style: $statusBarStyle")
            return null
        }
        val params = options.getMap("params")?.toHashMap().orEmpty()
        val overlay = options.getMap("overlay")
        val presentScrimColor = overlay?.getString("scrimColor")
        if (presentScrimColor != null && !SCRIM_COLOR.matches(presentScrimColor)) {
            onError("Invalid overlay scrim color: $presentScrimColor")
            return null
        }
        val presentBackdropTransition = overlay?.getBoolean("backdropTransition", true) ?: true
        val legacyContentTransition = overlay?.getBoolean("contentTransition", true) ?: true
        val presentEnter = overlay?.getMap("enter")
        val presentExit = overlay?.getMap("exit")
        val presentBackdropBlur = overlay?.getBoolean("backdropBlur", false) ?: false
        val presentAndroidPredictiveBackDown =
            overlay?.getBoolean("androidPredictiveBackDown", false) ?: false
        val presentDragDownToDismiss =
            overlay?.getBoolean("dragDownToDismiss", false) ?: false
        val presentOptions = PresentRouteOptions(
            scrimColor = presentScrimColor,
            backdropTransition = presentBackdropTransition,
            enterAnimation = PresentContentAnimationOptions(
                opacity = presentEnter?.getBoolean("opacity", false) ?: false,
                push = presentEnter?.getBoolean("push", legacyContentTransition)
                    ?: legacyContentTransition,
            ),
            exitAnimation = PresentContentAnimationOptions(
                opacity = presentExit?.getBoolean("opacity", false) ?: false,
                push = presentExit?.getBoolean("push", legacyContentTransition)
                    ?: legacyContentTransition,
            ),
            backdropBlur = presentBackdropBlur,
            androidPredictiveBackDown = presentAndroidPredictiveBackDown,
            dragDownToDismiss = presentDragDownToDismiss,
        )
        return ValidatedRoute(
            bundle = bundle,
            animation = animation,
            presentation = presentation,
            statusBarStyle = statusBarStyle,
            params = params,
            presentOptions = presentOptions,
        )
    }

    private fun routeResultToken(activity: Activity): String? =
        activity.intent?.getStringExtra(EXTRA_RESULT_TOKEN)

    /** Delivers the pending entry once, with its recorded result or none. */
    private fun completePendingResult(token: String) {
        val pending = pendingResults.remove(token) ?: return
        val result = pending.result
        val envelope = if (result != null) {
            RouteResultEnvelope.value(JSONObject(result))
        } else {
            RouteResultEnvelope.empty()
        }
        pending.resultCallback.invoke(envelope)
    }

    override fun onActivityDestroyed(activity: Activity) {
        // Route close: deliver its pending result (or the no-result envelope).
        routeResultToken(activity)?.let(::completePendingResult)
        // Opener death: the awaiting JS context is gone; drop its entries.
        val iterator = pendingResults.entries.iterator()
        while (iterator.hasNext()) {
            val entry = iterator.next().value
            if (entry.opener.get() === activity) {
                iterator.remove()
            }
        }
    }

    override fun onActivityCreated(activity: Activity, savedInstanceState: android.os.Bundle?) {}

    override fun onActivityStarted(activity: Activity) {}

    override fun onActivityResumed(activity: Activity) {}

    override fun onActivityPaused(activity: Activity) {}

    override fun onActivityStopped(activity: Activity) {}

    override fun onActivitySaveInstanceState(
        activity: Activity,
        outState: android.os.Bundle,
    ) {}

    companion object {
        private const val TAG = "AppRouteHandler"
        /** Correlates a launched route with its opener's pending callback. */
        internal const val EXTRA_RESULT_TOKEN = "lynx.route.result-token"
        private val BUNDLE_NAME = Regex("^[a-z0-9][a-z0-9-]*$")
        private val SCRIM_COLOR = Regex("^#[0-9a-fA-F]{8}$")
        private val NEXT_RESULT_TOKEN = AtomicLong()

        /** Invokes the callback with a JSON error envelope. */
        private fun Callback.invokeEnvelopeError(message: String) {
            invoke(RouteResultEnvelope.error(message))
        }
    }
}
